### PR TITLE
Basic Language Server

### PR DESCRIPTION
**Pull request by xackus**
Uses `std.zig.parse` to validate code as you type.
Work in progress, there may be some ugly code.
If you want to test this on vscode, use branch lsp of https://github.com/xackus/vscode-zig and change the hardcoded path to server.zig in extension.ts.

### CODE REVIEW

File: language-server/json_serialize.zig [hunk 1]
```diff
@@ -0,0 +1,405 @@
...
+    String: []const u8,
+};
+
+pub const Number = union(enum) {
+    Integer: i64,
+    Float: f64,
+};
+
+/// jsonStream must be a pointer to std.json.WriteStream
+pub fn serialize(value: var, jsonStream: var) !void {
```
Comment (daurnimator) [inline language-server/json_serialize.zig]: I have a WIP branch for new json stuff over at https://github.com/ziglang/zig/compare/master...daurnimator:auto-json
Will hopefully reopen the PR for it soon.

Comment (xackus) [reply language-server/json_serialize.zig]: Your branch looks nice, but I need the capabilities to not emit a struct field at all and tell the difference between `null` and not defined (see my `MaybeDefined`).

Comment (daurnimator) [reply language-server/json_serialize.zig]: 1. use a union of a `void` and the type?
2. use the custom dumper?

Comment (xackus) [reply language-server/json_serialize.zig]: - I don't want to write a custom dumper for every struct that has a field that may not be defined.
- I need the parser to distinguish between null and not defined.

Comment (daurnimator) [reply language-server/json_serialize.zig]: > I need the parser to distinguish between null and not defined.

It does: not defined values use their default values.

**Comment by andrewrk** [comment]
This looks like a good candidate for being a [third party project](https://github.com/ziglang/zig/wiki/Community-Projects), yeah?

**Comment by xackus** [comment]
I found #615 so I made a pull request. If you want, it can be a third party project.

**Comment by andrewrk** [comment]
#615 depends on the self-hosted compiler (#89) being done. It will have semantic analysis awareness, and integrate with the compiler itself. What could be useful right now would be filling the gap until such a thing is available, with a third-party project that uses stage1's `-fdump-analysis` to find type information.

**Comment by andrewrk** [comment]
@xackus this is a really exciting project! I invite you to start a separate source repository with a readme file that explains how to set it up, how to use it, etc, and I will be happy to link to it from official places such as the documentation and zig's readme.

The fact that all the files this modifies are in a new subdirectory called `language-server/` is a beautiful separation of concerns. It's a sign that you're doing it *correctly* that I'm asking you to maintain it as a separate project.

So I hope that it is clear that I am endorsing your project, while closing this pull request, because there are no stage1 compiler changes here.

**Comment by XVilka** [comment]
By the way, while this project should be in a separate repository, it still makes sense to put it in the https://github.com/ziglang/ organization, I think.

**Comment by JustinRyanH** [comment]
@xackus Hey Would you be interested in starting a project. I would love to help contribute to a ZigLS.

**Comment by xackus** [comment]
@JustinRyanH Sorry for taking long to respond.
I created a repo: https://github.com/xackus/zig-lsp.
I updated the code to zig 0.6.0, but it currently does not compile becasue it references `std.json.Value.dump`, which doesn't compile.
If you want to contribute, you can start by fixing the issue in zig std lib and adding a test that simply references the function to make sure it compiles.

**Comment by markfirmware** [comment]
@xackus @andrewrk @SuperAuguste @Vexu @alexnask 

It looks like zigtools/zls is taking off. Is there any guidance on where efforts should be made?

**Comment by SuperAuguste** [comment]
Hey @markfirmware, great question about Zig Language Servers! I don't think there's any official guidance on where efforts should be made as `xackus/zig-lsp` and `zigtools/zls` are both third-party projects! I believe you should contribute to any project you find interesting or think is a good investment!

**Comment by alexnask** [comment]
@markfirmware 

Ideally I would imagine stage2 could provide a langugae server and I would be happy to contribute to it once the codebase is stable enough and assuming that is within scope of stage2.  
No separate tool is ever going to be close to what we could do within the compiler itself unless it actually is a compiler (or at least an interpreter for the comptime subset of zig features)

**Comment by markfirmware** [comment]
Ok, thanks.

**Pull request closed without merging**